app/scripts/modules/oracle/serverGroup/configure/serverGroupConfiguration.service.js

```diff
@@ -0,0 +1,112 @@
+
+        return $q.all({
+            credentialsKeyedByAccount: accountService.getCredentialsKeyedByAccount(oracle),
+            networks: networkReader.listNetworksByProvider(oracle),
+            subnets: oraclebmcsSubnetReader.listSubnetsByProvider(oracle),
+            images: loadImages(),
+            availDomains: accountService.getAvailabilityZonesForAccountAndRegion(oracle, defaultCredentials, defaultRegion)
+        }).then(function(backingData) {
+            backingData.accounts = _.keys(backingData.credentialsKeyedByAccount);
+            backingData.filtered = {};
+            backingData.filtered.regions = [{name: 'us-phoenix-1'}];
+            backingData.filtered.availabilityDomains = _.map(backingData.availDomains, function(zone) {
+              return {name: zone};
+            });
+
+            backingData.filterSubnets = function() {
+              if(command.vpcId && command.availabilityDomain) {
+                return _.filter(backingData.subnets, {
+                  vcnId: command.vpcId,
+                  availabilityDomain: command.availabilityDomain
+                });
+              }
+              return backingData.subnets;
+            };
+
+            backingData.availabilityDomainOnChange = function() {
+              command.subnetId = null;
+              backingData.seclists = null;
+            };
+
+            backingData.vpcOnChange = function() {
+              command.subnetId = null;
+              backingData.seclists = null;
+            };
+
+            backingData.subnetOnChange = function() {
+              let subnet = _.find(backingData.subnets, {id : command.subnetId});
+
+              securityGroupReader.getAllSecurityGroups().then(function(allSecurityGroups) {
+                let mySecGroups = allSecurityGroups[command.account][oracle][command.region];
+                let secLists = [];
+                _.forEach(subnet.securityListIds, function(sid) {
+                  let sgRef = _.find(mySecGroups, {id: sid});
+                  securityGroupReader.getSecurityGroupDetails(
+                    command.application,
+                    command.account,
+                    oracle,
+                    command.region,
+                    command.vpcId,
+                    sgRef.name
+                  ).then(function(sgd) {
+                    secLists.push(sgd);
+                    backingData.seclists = secLists;
+                  });
+                });
+              });
+            };
+
+            backingData.filtered.images = backingData.images;
+            let shapesMap = {};
+            _.forEach(backingData.filtered.images, (image) => {
+              shapesMap[image.id] = getShapes(image);
+            });
+            backingData.filtered.shapes = shapesMap;
+            backingData.filtered.allShapes = _.uniqBy(_.flatMap(shapesMap, (v) => { return v; }), 'name');
```

Comment:
If you use `_.flatten` here instead of `_.flatMap`, you won't have to pass in an identity function.